plugins/logitech-scribe/fu-logitech-scribe-device.c

```diff
@@ -0,0 +1,603 @@
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 94, "device-write-blocks");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, "end-transfer");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 5, "uninit");
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* sending INIT. Retry if device is not in IDLE state to receive the file */
+	if (!fu_device_retry(device,
+			     fu_logitech_scribe_device_send_upd_init_cmd_cb,
+			     MAX_RETRIES,
+			     NULL,
+			     error)) {
+		g_prefix_error(error,
+			       "failed to write init transfer packet: please reboot the device: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* start transfer */
+	fu_byte_array_append_uint64(start_pkt, g_bytes_get_size(fw), G_LITTLE_ENDIAN);
+	if (!fu_logitech_scribe_device_send_upd_cmd(self, CMD_START_TRANSFER, start_pkt, error)) {
+		g_prefix_error(error, "failed to write start transfer packet: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* push each block to device */
+	if (!fu_logitech_scribe_device_write_fw(self, fw, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* end transfer */
+	base64hash = fu_logitech_scribe_device_compute_hash(fw);
+	fu_byte_array_append_uint32(end_pkt, 1, G_LITTLE_ENDIAN);   /* update */
+	fu_byte_array_append_uint32(end_pkt, 0, G_LITTLE_ENDIAN);   /* force */
+	fu_byte_array_append_uint32(end_pkt, MD5, G_LITTLE_ENDIAN); /* checksum type */
+	g_byte_array_append(end_pkt, (const guint8 *)base64hash, strlen(base64hash));
+	if (!fu_logitech_scribe_device_send_upd_cmd(self, CMD_END_TRANSFER, end_pkt, error)) {
+		g_prefix_error(error, "failed to write end transfer transfer packet: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* uninitialize */
+	/* no need to wait for ACK message, perhaps device reboot already in progress, ignore */
+	if (!fu_logitech_scribe_device_send_upd_cmd(self, CMD_UNINIT, NULL, &error_local)) {
+		g_debug("Failed to receive acknowledgment for uninitialize request, ignoring it");
+	}
+	fu_progress_step_done(progress);
+
+	/*
+	 * image file pushed. Device validates and uploads new image on inactive partition. Reboots
+	 * wait for RemoveDelay duration
+	 */
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_scribe_device_set_version(FuDevice *device, GError **error)
+{
+	FuLogitechScribeDevice *self = FU_LOGITECH_SCRIBE_DEVICE(device);
+	g_autofree gchar *fwversion_str = NULL;
+	guint32 fwversion = 0;
+	guint16 data_len = 0;
+	g_autofree guint8 *query_data = NULL;
+
+	/* query current device version */
+	if (!fu_logitech_scribe_device_query_data_size(self,
+						       kLogiUnitIdCameraVersion,
+						       kLogiCameraVersionSelector,
+						       &data_len,
+						       error))
+		return FALSE;
+	query_data = g_malloc0(data_len);
```

Comment:
i think you should check data_len is sane, e.g. < 1024 bytes or something.

Reply:
Fixed